### ISSUE TITLE
Automatically set up gitHooks

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#581]
In GitLab by @mcastril on Sep 7, 2020, 18:02

It would be really helpful that Autosubmit automatically modifies the cloned project git config so the .githooks folder, in case it exists, is set as default githooks directory:

`git config core.hooksPath .githooks`.

Githooks are automated triggers, stored in the Git configuration, and because of that they cannot be control versioned. With this approach we would allow projects to include and version their own hooks, useful to do automated tasks after certain git operations.

As explained here: https://earth.bsc.es/gitlab/es/auto-ecearth3/-/issues/1378

Hopefully, it could also be used to move the NEMO clone script ( https://earth.bsc.es/gitlab/svn/nemo-mirror/snippets/55 ) to a githook-based thing.

**Comment by kinow** [comment]
In GitLab by @mcastril on Sep 22, 2020, 15:59

mentioned in issue auto-ecearth3#1378

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Oct 1, 2020, 11:55

Will be released on next update

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 1, 2020, 12:17

Nice!

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 2, 2020, 13:17

@sparonuz may actually be a good tester for this, as she's already using githooks.

**Comment by kinow** [comment]
In GitLab by @sparonuz on Oct 2, 2020, 14:38

Hi Miguel, actually I am not using autosubmit that much right now, an it is a specific autosubmit feature as I understand. Tell me if I can be of any help!

**Comment by kinow** [comment]
In GitLab by @mcastril on Oct 2, 2020, 16:14

Ah ok, I thought you were using it in your Auto-Performance workflow. Yes, it is Autosubmit specific. So well, just in case you use it please tell if it works.

**Comment by kinow** [comment]
In GitLab by @sparonuz on Oct 2, 2020, 16:41

We were planning to use it, but for the moment was not a priority, so I haven't followed along with the implementation.
When I start again developing this part for sure I'll test the hooks!
Thanks

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 16, 2021, 11:18

@dbeltrankyl can you check in which specific moment the githook is set up?

We would be interested in having a githook active at the moment that the submodules are cloned.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jan 20, 2022, 11:02

After some tests I detected some bugs, now it is ready to be used.

```
(venv27) dbeltran@kyl:~/experiments/a03g/proj/githooktest$ touch hola
(venv27) dbeltran@kyl:~/experiments/a03g/proj/githooktest$ git commit -am "hola"
```

```
2 files changed, 0 insertions(+), 0 deletions(-)
 mode change 100644 => 100755 .githooks/.gitkeep
 mode change 100644 => 100755 .githooks/pre-commit
(venv27) dbeltran@kyl:~/experiments/a03g/proj/githooktest$ ls
README.md  commit_world.out  hola  remote_setup.sh
```

Pre-commit file contains: ( slightly modified from git example, in the end, it only creates the file commit_world.out) 

```
#!/bin/sh
#
# An example hook script to verify what is about to be committed.
# Called by "git commit" with no arguments.  The hook should
# exit with non-zero status after issuing an appropriate message if
# it wants to stop the commit.
#
# To enable this hook, rename this file to "pre-commit".

if git rev-parse --verify HEAD >/dev/null 2>&1
then
        against=HEAD
else
        # Initial commit: diff against an empty tree object
        against=$(git hash-object -t tree /dev/null)
fi

# If you want to allow non-ASCII filenames set this variable to true.
allownonascii=$(git config --type=bool hooks.allownonascii)

# Redirect output to stderr.
exec 1>&2

# Cross platform projects tend to avoid non-ASCII filenames; prevent
# them from being added to the repository. We exploit the fact that the
# printable range starts at the space character and ends with tilde.
if [ "$allownonascii" != "true" ] &&
        # Note that the use of brackets around a tr range is ok here, (it's
        # even required, for portability to Solaris 10's /usr/bin/tr), since
        # the square bracket bytes happen to fall in the designated range.
        test $(git diff --cached --name-only --diff-filter=A -z $against |
          LC_ALL=C tr -d '[ -~]\0' | wc -c) != 0
then
        cat <<\EOF
Error: Attempt to add a non-ASCII file name.

This can cause problems if you want to work with people on other platforms.

To be portable it is advisable to rename the file.

If you know what you are doing you can disable this check using:

  git config hooks.allownonascii true
EOF
        exit 1
fi

touch commit_world.out
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 27, 2022, 12:03

Thank you @dbeltrankyl !

@eferre1 I will assign you the "gitHook-task" in Auto-EC-Earth so can you can test if it's ready for our use case.

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 1, 2022, 12:24

I've tested this and it works (sets up the `.githooks` as the new githooks path) but not for our case, since the change of the path is done after the submodules are initialized. I'm guessing that AS runs `git clone --recurse-submodules` and downloads the repositories all in the same step instead of doing it in two steps with the submodules initialization after the main repository clone.

I don't know if that can be easily changed or not, if it is possible it would be perfect, if not I'll look for a way of making it work as it is.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 1, 2022, 12:28

Hello @eferre1 ,

Autosubmit is already doing it before the submodules init, so I'm not sure why it is not working

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 1, 2022, 12:32

edit; The version is 3.14.0b

The order ( there are some variants, depending on the behaviour indicated) but the main workflow is the next:

1) git clone... ; git checkout ; ...
2) git config core.hooksPath .githooks; 
3) git submodule init;
4) git submodule update;

Do you see something wrong in this order?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 1, 2022, 12:35

Oh, maybe the execute permissions must be also set before the submodule cloning

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 1, 2022, 12:38

I was running the AS create (after removing the proj folder) and it didn't seem to be setting the path before cloning the modules, I'm also not sure if maybe it was doing it and it is that the submodule init and update doesn't trigger the selected hook. I will check if that happens manually to be sure.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 1, 2022, 12:40

The are two things to check:

* It seems that the `.git/config` is not changed until the submodules are cloned.
* We would need the `.git/config` to be set before the `git checkout` to use it as a trigger.

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 1, 2022, 17:05

Hi @dbeltrankyl, if you want to check, at `/esarchive/autosubmit/a3og/` you can see that it was running the `--refresh` command and it has 2 proj folders, and the one were it was cloning the repository (`proj/auto-ecearth3`) has already 2 submodules cloned there (`ece2cmor` and `ece3-postproc`) but on the `.git/config` the `core.hooksPath` has not been set yet. (I stopped the refresh with ctrl+C to check it out).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 2, 2022, 10:34

Hello @eferre1 ,

I still need to upload the changes to checkout, but the issue that you encountered is that you're using the default behavior of autosubmit, which is using --recursive. 

To avoid this default behavior, you can set the list of submodules by using the parameter ```PROJECT_SUBMODULES```, otherwise autosubmit is unable to load them one by one.

```
    def get_submodules_list(self):
        """
        Returns submodules list from experiment's config file
        Default is --recursive
        :return: submodules to load
        :rtype: list
        """
        return ' '.join(self._exp_parser.get_option('git', 'PROJECT_SUBMODULES', '').split()).split()
```


So, you have to list in the git section of expdef.conf, the submodules that you want to load

The syntaxis is:` PROJECT_SUBMODULES = sub1 sub2 sub3 `...

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 2, 2022, 10:37

I'll add this parameter for future new experiments as empty.

edit: nevermind, it is already done

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 2, 2022, 10:40

Ok, then unless using the `PROJECT_SUBMODULES`, then AS runs `git clone --recurse-submodules`, and if it is defined then it is when runs in a 2 step procedure with the githooks in the middle.

That parameter has to be defined in the expdef, right? I haven't seen it in the auto-ecearth3 test_cases, so I guess it was removed in the past since it wasn't being used, but we'll have to re-add it if we want to use the githooks feature.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 2, 2022, 10:43

Yes, these are the parameters of [git] when the experiment is created from scratch.

```
# If PROJECT_TYPE is not git, no need to change
[git]
# Repository URL  STRING = 'https://github.com/torvalds/linux.git'
PROJECT_ORIGIN =
# Select branch or tag, STRING, default = 'master', help = {'master' (default), 'develop', 'v3.1b', ...}
PROJECT_BRANCH =
# type = STRING, default = leave empty, help = if model branch is a TAG leave empty
PROJECT_COMMIT =
# type = STRING, default = leave empty and will load all submodules, help = loadThisSubmodule alsoloadthis anotherLoad ...
PROJECT_SUBMODULES =
# type = STRING, default = leave empty and will perform a full clone, help = true,false
FETCH_SINGLE_BRANCH = true
# type = STRING, default = leave empty and will perform a clone on local machine, help = Path to root of remote folder
#REMOTE_CLONE_ROOT = /gpfs/scratch/archive/bsc32
```

From your experiment, FETCH_SINGLE_BATCH and PROJECT_SUBMODULES are missing. About FETCH_SINGLE_BATCH, this parameter was added to save storage so is recommended to be used

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 2, 2022, 10:53

Ok, thanks, I added both parameters and now it is running as you explained above, but there is a typo and it is failing because of that:

```
Cloning 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git into /esarchive/autosubmit/a3og/proj
/bin/sh: -c: line 0: syntax error near unexpected token `;;'
/bin/sh: -c: line 0: `cd /esarchive/autosubmit/a3og/proj;  git clone --single-branch -b 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git auto-ecearth3;; cd auto-ecearth3 ; git config core.hooksPath .githooks; cd .. ;  cd auto-ecearth3; git submodule init; git submodule update  sources; git submodule update  ece2cmor; git submodule update  ece3-postproc; git submodule update  es-esmvaltool; git submodule update  test_cases; git submodule update  utils;'
 [CRITICAL] Can not clone 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git into /esarchive/autosubmit/a3og/proj [eCode=7065]
```

I'll wait until the fix. 

Also for the purposes of the hook running in between the clone and the submodules initialization we would need that `git checkout`. And I think it is missing a `../` in the hooks setting path (I was using `git config core.hooksPath ./.githooks` during my manual tests).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 2, 2022, 11:00

Oh, Didn't notice the ;; error. Thanks. 

Anyway, I've refactored it a bit and changed the checkout place. Also, the permissions to the githook folder will be set before the checkout and the submodules.

I'll update the version late today, will mention you here whenever it is ready for the modules.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 2, 2022, 16:38

> From your experiment, FETCH_SINGLE_BATCH and PROJECT_SUBMODULES are missing. About FETCH_SINGLE_BATCH, this parameter was added to save storage so is recommended to be used

Thank you @dbeltrankyl. In fact, I think that these parameter was done on purpose for Auto-EC-Earth3, don't know in which moment disappeared from the tests

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 2, 2022, 16:52

I've had a look, and those 2 parameters never made it into my tests and neither are they on the release tests, so maybe they were lost during the auto-ecearth3 refactor since weren't being used in the templates?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 3, 2022, 13:30

Hello, It is updated on the modules, yesterday I was unable to run tests due platform issues, but I have one last question.

@mcastril  You mentioned setting the githook before the checkout. However, the checkout command is only used when there is specified a commit in `PROJECT_COMMIT`. Otherwise, it just will do git clone -b <branch> without checkout as always did Autosubmit.

Do you see any inconvenience on there?

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 3, 2022, 14:33

It could be that the githook is not triggered because we are using the commit event. There is no clone githook.

@eferre1 will check that.

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 4, 2022, 12:22

I'm not sure if it will work without the checkout, because the hook that we were thinking of is the `post-checkout` (from the [documentation](https://git-scm.com/docs/githooks)). I'll test it and let you know, but maybe we will need that checkout to be called empty if there is no `PROJECT_COMMIT` defined for the hook to work.

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 8, 2022, 08:49

Hi @dbeltrankyl, I'm testing it with the `module load autosubmit/3.14.0b-foss-2015a-Python-2.7.9` version, but it doesn't seem to be setting up the hooks. I'm using the next parameters in the git section:

```
[git]
# Repository URL  STRING = 'https://github.com/torvalds/linux.git'
PROJECT_ORIGIN = https://earth.bsc.es/gitlab/es/auto-ecearth3.git
# Select branch or tag, STRING, default = 'master', help = {'master' (default), 'develop', 'v3.1b', ...}
PROJECT_BRANCH = 1378-githooks_setup
# type = STRING, default = leave empty, help = if model branch is a TAG leave empty
PROJECT_COMMIT =
# type = STRING, default = leave empty and will load all submodules, help = loadThisSubmodule alsoloadthis anotherLoad ...
PROJECT_SUBMODULES = sources ece2cmor ece3-postproc es-esmvaltool test_cases utils
# type = STRING, default = leave empty and will perform a full clone, help = true,false
FETCH_SINGLE_BRANCH = true
```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 8, 2022, 15:11

Just as a note, if you add the `git checkout` without any argument, what the checkout does according to the documentation is: 
>If no pathspec was given, git checkout will also update HEAD to set the specified branch as the current branch.

So, when doing the 2 step clone it could be added in a way that does the empty checkout or the checkout of the commit if available. That way, it will activate the hook.

Adding a bit more complexity, it could be checked if the path specified for the hooks exists to avoid doing the empty checkout if it doesn't.

Testing manually, but following the AS sequence of commands, this would be the looks of it for the auto-ecearth3 create/refresh:

<details><summary>AS expected create/refresh results</summary>

```
eferre1@ bscearth358:/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test> git clone --single-branch -b 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git auto-ecearth3
Cloning into 'auto-ecearth3'...
remote: Enumerating objects: 19661, done.
remote: Counting objects: 100% (5920/5920), done.
remote: Compressing objects: 100% (1175/1175), done.
remote: Total 19661 (delta 5614), reused 4770 (delta 4744), pack-reused 13741
Receiving objects: 100% (19661/19661), 18.93 MiB | 3.83 MiB/s, done.
Resolving deltas: 100% (13199/13199), done.
Checking out files: 100% (378/378), done.
eferre1@ bscearth358:/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test> cd auto-ecearth3/
eferre1@ bscearth358:/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3> git config core.hooksPath ./.githooks
eferre1@ bscearth358:/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3> git checkout
Your branch is up-to-date with 'origin/1378-githooks_setup'.
Running post-checkout githook to configure the SVN repositories for sources
Submodule 'sources' (https://earth.bsc.es/gitlab/svn/ecearth-mirror.git) registered for path 'sources'
Cloning into '/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3/sources'...
Submodule path 'sources': checked out 'ceb1f5b95ced1374f190d2a9f9b833a74627c5d8'
SVN repository set up, githooks restored to .git/hooks
eferre1@ bscearth358:/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3> git submodule init
Submodule 'ece2cmor' (https://earth.bsc.es/gitlab/es/ece2cmor.git) registered for path 'ece2cmor'
Submodule 'ece3-postproc' (https://earth.bsc.es/gitlab/es/ece3-postproc.git) registered for path 'ece3-postproc'
Submodule 'es-esmvaltool' (https://earth.bsc.es/gitlab/es/es-esmvaltool.git) registered for path 'es-esmvaltool'
Submodule 'test_cases' (https://earth.bsc.es/gitlab/es/auto-ecearth-conf-files.git) registered for path 'test_cases'
Submodule 'utils' (https://earth.bsc.es/gitlab/es/auto-ecearth-utils.git) registered for path 'utils'
eferre1@ bscearth358:/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3> git submodule update
Cloning into '/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3/ece2cmor'...
Cloning into '/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3/ece3-postproc'...
Cloning into '/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3/es-esmvaltool'...
Cloning into '/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3/test_cases'...
Cloning into '/esarchive/scratch/bsc32627/git/auto-ecearth3_githooks/test/auto-ecearth3/utils'...
Submodule path 'ece2cmor': checked out '3aeaaf08bcf171b36c76caf59a1d5c842bec2798'
Submodule path 'ece3-postproc': checked out '0715f77dfdd2bc02baa37bdbb1132ca8448c0fde'
Submodule path 'es-esmvaltool': checked out 'c30559401e0a2a5cae89d7456a25ecb6cdefca1a'
Submodule path 'test_cases': checked out '941199f851df317e13bbc40f2f03d62e08bc3070'
Submodule path 'utils': checked out 'c293286161cfdb489e67b74b37a590e844460427'
```

</details>

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 8, 2022, 16:39

Thanks @eferre1 

But it seems that currently the .githooks are not properly configured after the first `git clone`. Can you take a look @dbeltrankyl ?

> Adding a bit more complexity, it could be checked if the path specified for the hooks exists to avoid doing the empty checkout if it doesn't.

We need to do this because the `git checkout` sometimes take time, so I prefer to do it only if necessary.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 9, 2022, 15:07

Hello, 

> But it seems that currently the .githooks are not properly configured after the first `git clone`. Can you take a look @dbeltrankyl ?

Fixed

> Adding a bit more complexity, it could be checked if the path specified for the hooks exists to avoid doing the empty checkout if it doesn't.

Added


I update the module 3.14.0b, the githook is now correctly running however I don't see the same output as @eferre1  described:

```
dbeltran@ bscearth357:~> module load autosubmit/3.14.0b-foss-2015a-Python-2.7.9 
dbeltran@ bscearth357:~> autosubmit expid --copy a3og -H "marenostrum4" -d "githook"
The new experiment "a4bk" has been registered.
dbeltran@ bscearth357:/esarchive/autosubmit/a4bk/conf> autosubmit -lc "DEBUG" refresh a4bk
(clone part)
....
Githook + Checkout and Submodules: cd /esarchive/autosubmit/a4bk/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
Running post-checkout githook
/esarchive/autosubmit/a4bk/proj/auto-ecearth3
168c9955e7032511af6fd3ea9d03eaae13ce9113
168c9955e7032511af6fd3ea9d03eaae13ce9113
1
Doing nothing
Submodule 'ece2cmor' (https://earth.bsc.es/gitlab/es/ece2cmor.git) registered for path 'ece2cmor'
Submodule 'ece3-postproc' (https://earth.bsc.es/gitlab/es/ece3-postproc.git) registered for path 'ece3-postproc'
Submodule 'es-esmvaltool' (https://earth.bsc.es/gitlab/es/es-esmvaltool.git) registered for path 'es-esmvaltool'
Submodule 'sources' (https://earth.bsc.es/gitlab/svn/ecearth-mirror.git) registered for path 'sources'
Submodule 'test_cases' (https://earth.bsc.es/gitlab/es/auto-ecearth-conf-files.git) registered for path 'test_cases'
Submodule 'utils' (https://earth.bsc.es/gitlab/es/auto-ecearth-utils.git) registered for path 'utils'
Cloning into '/esarchive/autosubmit/a4bk/proj/auto-ecearth3/test_cases'...
Project folder updated
```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 9, 2022, 15:13

Yes, the hook on the auto-ecearth3 branch is still not updated to what it should be, that was the first version with just some messages to check that it was working. I was manually testing it and will update it to test it with AS now too.

Thanks.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 9, 2022, 15:13

Ah, okay.
 thanks

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 15, 2022, 09:22

Hi @dbeltrankyl, now you can test it if you want since the githook it is updated on the branch. I'm getting some weird results:

- autosubmit refresh:
```
eferre1@ bscesautosubmit02:~> autosubmit -lc "DEBUG" refresh a3og
Locale C.utf8 is not found, using 'C' as fallback

Checking configuration files...
expdef_a3og.conf OK
platforms_a3og.conf OK
jobs_a3og.conf OK
wrappers OK
autosubmit_a3og.conf OK
Configuration files OK

Custom directives from platform.conf: None
Custom directives from platform.conf: None
Custom directives from platform.conf: None
Custom directives from platform.conf: None
Making a backup of your current proj folder at /esarchive/autosubmit/a3og/proj_1644912156
The project folder /esarchive/autosubmit/a3og/proj has been created.
Cloning 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git into /esarchive/autosubmit/a3og/proj
Clone command: cd /esarchive/autosubmit/a3og/proj;  git clone --single-branch -b 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git auto-ecearth3;
Cloning into 'auto-ecearth3'...
Username for 'https://earth.bsc.es': eferre1
Password for 'https://eferre1@ earth.bsc.es': 
remote: Enumerating objects: 19665, done.
remote: Counting objects: 100% (5924/5924), done.
remote: Compressing objects: 100% (1178/1178), done.
remote: Total 19665 (delta 5616), reused 4768 (delta 4744), pack-reused 13741
Receiving objects: 100% (19665/19665), 18.93 MiB | 4.29 MiB/s, done.
Resolving deltas: 100% (13201/13201), done.
Checking connectivity... done.
Checking out files: 100% (378/378), done.
Githook + Checkout and Submodules: cd /esarchive/autosubmit/a3og/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
Cloning into 'test_cases'...
Username for 'https://earth.bsc.es': eferre1
Password for 'https://eferre1@ earth.bsc.es': 
remote: Enumerating objects: 7884, done.
remote: Counting objects: 100% (1286/1286), done.
remote: Compressing objects: 100% (520/520), done.
remote: Total 7884 (delta 767), reused 781 (delta 401), pack-reused 6598
Receiving objects: 100% (7884/7884), 897.80 KiB | 123.00 KiB/s, done.
Resolving deltas: 100% (4866/4866), done.
Checking connectivity... done.
Removing backup...
Project folder updated
```

- manually running `cd /esarchive/autosubmit/a3og/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;`

```
eferre1@ bscearth358:/esarchive/autosubmit/a3og> cd /esarchive/autosubmit/a3og/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
Your branch is up-to-date with 'origin/1378-githooks_setup'.
Running post-checkout githook to configure the SVN repositories for sources
Cloning into '/esarchive/autosubmit/a3og/proj/auto-ecearth3/sources'...
Submodule path 'sources': checked out 'ceb1f5b95ced1374f190d2a9f9b833a74627c5d8'
SVN repository set up, githooks restored to .git/hooks
```

So, it seems like it should be running fine, but it is not triggering the githook for me, even though running the same commands that autosubmit uses then it does.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 15, 2022, 10:00

Hello @eferre1 ,

The last version that I tested worked. But this version didn't seem to work.

I can't replicate your success even if I do it manually:

Autosubmit control + c in the second step result:

```
Githook + Checkout and Submodules: cd /esarchive/autosubmit/a4bk/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
Cloning into 'test_cases'...
remote: Enumerating objects: 7884, done.
remote: Counting objects: 100% (1286/1286), done.
remote: Compressing objects: 100% (520/520), done.
^CTrace:  objects:   3% (237/7884), 44.01 KiB | 72.00 KiB/s   
Submodule test_cases has a wrong configuration
Some Submodule failures have been detected. Backup /esarchive/autosubmit/a4bk/proj_1644914769 will not be removed.
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj> cd /esarchive/autosubmit/a4bk/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
D       test_cases
Your branch is up-to-date with 'origin/1378-githooks_setup'.
Cloning into 'test_cases'...
remote: Enumerating objects: 7884, done.
remote: Counting objects: 100% (1286/1286), done.
remote: Compressing objects: 100% (520/520), done.
remote: Total 7884 (delta 767), reused 781 (delta 401), pack-reused 6598
Receiving objects: 100% (7884/7884), 897.80 KiB | 131.00 KiB/s, done.
Resolving deltas: 100% (4866/4866), done.
Checking connectivity... done.
```

```
Githook + Checkout and Submodules: cd /esarchive/autosubmit/a4bk/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
Cloning into 'test_cases'...
remote: Enumerating objects: 7884, done.
remote: Counting objects: 100% (1286/1286), done.
remote: Compressing objects: 100% (520/520), done.
^C

dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj> cd /esarchive/autosubmit/a4bk/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
D       test_cases
Your branch is up-to-date with 'origin/1378-githooks_setup'.
Cloning into 'test_cases'...
remote: Enumerating objects: 7884, done.
remote: Counting objects: 100% (1286/1286), done.
remote: Compressing objects: 100% (520/520), done.
remote: Total 7884 (delta 767), reused 781 (delta 401), pack-reused 6598
Receiving objects: 100% (7884/7884), 897.80 KiB | 131.00 KiB/s, done.
Resolving deltas: 100% (4866/4866), done.
Checking connectivity... done.
```

Manually:

Step 1:

```
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj> cd /esarchive/autosubmit/a4bk/proj;  git clone --single-branch -b 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git auto-ecearth3;
Cloning into 'auto-ecearth3'...
remote: Enumerating objects: 19665, done.
remote: Counting objects: 100% (5924/5924), done.
remote: Compressing objects: 100% (1178/1178), done.
remote: Total 19665 (delta 5616), reused 4768 (delta 4744), pack-reused 13741
Receiving objects: 100% (19665/19665), 18.93 MiB | 3.95 MiB/s, done.
Resolving deltas: 100% (13201/13201), done.
Checking connectivity... done.
Checking out files: 100% (378/378), done.
```

Permissions seem okay, don't need chmod

step 2:

```
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj>  cd /esarchive/autosubmit/a4bk/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
Your branch is up-to-date with 'origin/1378-githooks_setup'.
Submodule 'ece2cmor' (https://earth.bsc.es/gitlab/es/ece2cmor.git) registered for path 'ece2cmor'
Submodule 'ece3-postproc' (https://earth.bsc.es/gitlab/es/ece3-postproc.git) registered for path 'ece3-postproc'
Submodule 'es-esmvaltool' (https://earth.bsc.es/gitlab/es/es-esmvaltool.git) registered for path 'es-esmvaltool'
Submodule 'sources' (https://earth.bsc.es/gitlab/svn/ecearth-mirror.git) registered for path 'sources'
Submodule 'test_cases' (https://earth.bsc.es/gitlab/es/auto-ecearth-conf-files.git) registered for path 'test_cases'
Submodule 'utils' (https://earth.bsc.es/gitlab/es/auto-ecearth-utils.git) registered for path 'utils'
Cloning into 'test_cases'...
remote: Enumerating objects: 7884, done.
remote: Counting objects: 100% (1286/1286), done.
remote: Compressing objects: 100% (520/520), done.
remote: Total 7884 (delta 767), reused 781 (delta 401), pack-reused 6598
Receiving objects: 100% (7884/7884), 897.80 KiB | 128.00 KiB/s, done.
Resolving deltas: 100% (4866/4866), done.
Checking connectivity... done.
Submodule path 'test_cases': checked out '941199f851df317e13bbc40f2f03d62e08bc3070'
```

I don't see any of the echos of the post-checkout

edit: @eferre1 Couldn't be that it is just that the logs are not being shown but the command is working? what this file is supposed to do? how can I check it worked without relying in the logs?

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 15, 2022, 10:13

At the end of the hook, the idea is that the sources submodule will be copied and the svn repositories set up you can check that in the file `.git/modules/sources/config`, and also I unset the githooks path (in `.git/config `) to avoid that a new checkout runs it again. You can check what the results should look like in `/esarchive/autosubmit/a3og/proj_example/auto-ecearth3`.

I was thinking that maybe the issue is that autosubmit isn't showing this inner echos and also since it needs to ask for the user and password to clone the sources submodule, but is in the script inside, autosubmit is not asking for it and that makes it fail?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 15, 2022, 10:23

>I was thinking that maybe the issue is that autosubmit isn't showing this inner echos and also since it needs to ask for the user and password to clone the sources submodule, but is in the script inside, autosubmit is not asking for it and that makes it fail?

I already checked it because, in my case, autosubmit never ask me for a password so it doesn't seem to be the issue.


> can check that in the file `.git/modules/sources/config`

It seems that I don't have this file, could it be the issue?
```
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj/auto-ecearth3> ll -hart /esarchive/autosubmit/a3og/proj_example/auto-ecearth3/.git/modules/sources/config
-rw-r--r-- 1 eferre1 Earth 1.2K Feb 15 10:09 /esarchive/autosubmit/a3og/proj_example/auto-ecearth3/.git/modules/sources/config
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj/auto-ecearth3> ll -hart .git/modules/sources/config
ls: cannot access .git/modules/sources/config: No such file or directory
```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 15, 2022, 10:27

No, since that file will be created by the submodule update that is run into the hook, that just means that it is not running it, if you check the `auto-ecearth3/sources/` folder and it is empty it will confirm this. That would mean that the hook is not running and it is failing before starting the script. Or that it isn't getting triggered by some reason.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 15, 2022, 10:32

But I don't have it even if I run the command without autosubmit

```
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj> cd /esarchive/autosubmit/a4bk/proj;  git clone --single-branch -b 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git auto-ecearth3;
Cloning into 'auto-ecearth3'...
remote: Enumerating objects: 19665, done.
remote: Counting objects: 100% (5924/5924), done.
remote: Compressing objects: 100% (1178/1178), done.
remote: Total 19665 (delta 5616), reused 4768 (delta 4744), pack-reused 13741
Receiving objects: 100% (19665/19665), 18.93 MiB | 4.31 MiB/s, done.
Resolving deltas: 100% (13201/13201), done.
Checking connectivity... done.
Checking out files: 100% (378/378), done.
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj> cd /esarchive/autosubmit/a4bk/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
Your branch is up-to-date with 'origin/1378-githooks_setup'.
Submodule 'ece2cmor' (https://earth.bsc.es/gitlab/es/ece2cmor.git) registered for path 'ece2cmor'
Submodule 'ece3-postproc' (https://earth.bsc.es/gitlab/es/ece3-postproc.git) registered for path 'ece3-postproc'
Submodule 'es-esmvaltool' (https://earth.bsc.es/gitlab/es/es-esmvaltool.git) registered for path 'es-esmvaltool'
Submodule 'sources' (https://earth.bsc.es/gitlab/svn/ecearth-mirror.git) registered for path 'sources'
Submodule 'test_cases' (https://earth.bsc.es/gitlab/es/auto-ecearth-conf-files.git) registered for path 'test_cases'
Submodule 'utils' (https://earth.bsc.es/gitlab/es/auto-ecearth-utils.git) registered for path 'utils'
Cloning into 'test_cases'...
remote: Enumerating objects: 7884, done.
remote: Counting objects: 100% (1286/1286), done.
remote: Compressing objects: 100% (520/520), done.
remote: Total 7884 (delta 767), reused 781 (delta 401), pack-reused 6598
Receiving objects: 100% (7884/7884), 897.80 KiB | 128.00 KiB/s, done.
Resolving deltas: 100% (4866/4866), done.
Checking connectivity... done.
Submodule path 'test_cases': checked out '941199f851df317e13bbc40f2f03d62e08bc3070'
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj/auto-ecearth3> ls
ece2cmor  ece3-postproc  es-esmvaltool  fx_files  outclass  platforms  plugins  README.md  sources  templates  test_cases  utils
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj/auto-ecearth3> ll -hart .git/modules/sources/config
ls: cannot access .git/modules/sources/config: No such file or directory
```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 15, 2022, 10:34

If you run 
```
cd /esarchive/autosubmit/a4bk/proj/auto-ecearth3
git checkout
```
now that the githook path is set, it doesn't trigger it?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 15, 2022, 10:35

no

```
dbeltran@ bscesautosubmit01:/esarchive/autosubmit/a4bk/proj/auto-ecearth3> git checkout
Your branch is up-to-date with 'origin/1378-githooks_setup'.

```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 15, 2022, 10:38

Well, that is weird, it should trigger it, and I've looked at it and you have all the permissions set, so I'm not sure why it is not working.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 15, 2022, 10:44

This is maybe the kind of not obvious thing that could be sorted out in a call. Maybe you can work together for one hour and get it through, I am sure that it will be more efficient.

> At the end of the hook, the idea is that the sources submodule will be copied and the svn repositories set up you can check that in the file `.git/modules/sources/config`, and also I unset the githooks path (in `.git/config `) to avoid that a new checkout runs it again.

@eferre1 that could prevent us from adding additional hooks. Maybe you can just add a check inside the hook to not clone again the model.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 15, 2022, 12:43

I am going to try the manual way to see if I get it working on my side.

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 15, 2022, 13:21

Sure, I could set it to change the script name so it won't be a hook anymore, a `mv .githooks/post-checkout .githooks/post-checkout_completed` or something like that would do the trick I guess.

What is weird is that for @dbeltrankyl, the hook is not activating even when running the checkout manually, since he has the hook properly set up and the file has the execution permisions as it is required. I would say that there is another different problem there, but yes, we could meet and have a look at it together.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 15, 2022, 14:55

I tried the manual way and it works as expected:

```
git clone --single-branch -b 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git auto-ecearth3;
Cloning into 'auto-ecearth3'...
remote: Enumerating objects: 19665, done.
remote: Counting objects: 100% (5924/5924), done.
remote: Compressing objects: 100% (1178/1178), done.
remote: Total 19665 (delta 5616), reused 4768 (delta 4744), pack-reused 13741
Receiving objects: 100% (19665/19665), 18.93 MiB | 27.70 MiB/s, done.
Resolving deltas: 100% (13201/13201), done.
```

```
cd auto-ecearth3/
git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
Your branch is up-to-date with 'origin/1378-githooks_setup'.
Running post-checkout githook to configure the SVN repositories for sources
Submodule 'sources' (https://earth.bsc.es/gitlab/svn/ecearth-mirror.git) registered for path 'sources'
Cloning into '/scratch/mcastril/auto-ecearth3/sources'...
Submodule path 'sources': checked out 'ceb1f5b95ced1374f190d2a9f9b833a74627c5d8'
SVN repository set up, githooks restored to .git/hooks
Submodule 'ece2cmor' (https://earth.bsc.es/gitlab/es/ece2cmor.git) registered for path 'ece2cmor'
Submodule 'ece3-postproc' (https://earth.bsc.es/gitlab/es/ece3-postproc.git) registered for path 'ece3-postproc'
Submodule 'es-esmvaltool' (https://earth.bsc.es/gitlab/es/es-esmvaltool.git) registered for path 'es-esmvaltool'
Submodule 'test_cases' (https://earth.bsc.es/gitlab/es/auto-ecearth-conf-files.git) registered for path 'test_cases'
Submodule 'utils' (https://earth.bsc.es/gitlab/es/auto-ecearth-utils.git) registered for path 'utils'
Cloning into '/scratch/mcastril/auto-ecearth3/test_cases'...
Submodule path 'test_cases': checked out '941199f851df317e13bbc40f2f03d62e08bc3070'
```

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 15, 2022, 15:01

With export GIT_TRACE=true and export GIT_CURL_VERBOSE=true (maybe the second is not needed):

```
 git config core.hooksPath ./.githooks ; git checkout;  git submodule init; git submodule update test_cases;
14:59:59.483988 git.c:322               trace: built-in: git 'config' 'core.hooksPath' './.githooks'
14:59:59.491816 git.c:322               trace: built-in: git 'checkout'
Your branch is up-to-date with 'origin/1378-githooks_setup'.
14:59:59.522487 run-command.c:626       trace: run_command: '.githooks/post-checkout' 'd884c7820c4a748c7454dd842b7323fe9ebb173e' 'd884c7820c4a748c7454dd842b7323fe9ebb173e' '1'
Running post-checkout githook to configure the SVN repositories for sources
14:59:59.525927 git.c:547               trace: exec: 'git-submodule' 'init' 'sources/'
14:59:59.525947 run-command.c:626       trace: run_command: 'git-submodule' 'init' 'sources/'
```
Maybe you can try that @dbeltrankyl

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Feb 15, 2022, 15:38

IT still doesn't do anything for me 

```
15:31:16.768549 git.c:349               trace: built-in: git 'config' 'core.hooksPath' './.githooks'
15:31:17.149107 git.c:349               trace: built-in: git 'checkout'
Your branch is up-to-date with 'origin/1378-githooks_setup'.
15:31:17.800145 git.c:554               trace: exec: 'git-submodule' 'init'
15:31:17.800272 run-command.c:341       trace: run_command: 'git-submodule' 'init'
```


Not sure why

@mcastril , Couldn't it be any related to permission for auto-ecearth project?

**Comment by kinow** [comment]
In GitLab by @eferre1 on Feb 15, 2022, 16:03

I don't know if this could be the issue causing the errors, but having this line `#!/bin/bash` instead of `#!/usr/bin/env bash` at the beginning of the script could influence the execution? From what I've read it is better to avoid the second option since it could change the version used, and the first option would use the same for each user, so it shouldn't be the case, but at this point I have no clue about why is working for @mcastril  and myself and not for @dbeltrankyl or with the autosubmit commands.

PS: I don't think it is this, since with the trace doesn't show anything about running the post-checkout script, but just to be sure in case that it could be a source of errors in the future.

**Comment by kinow** [comment]
In GitLab by @mcastril on Feb 15, 2022, 16:33

> PS: I don't think it is this, since with the trace doesn't show anything about running the post-checkout script, but just to be sure in case that it could be a source of errors in the future.

I think the same, but we should try any possibility.

> @mcastril , Couldn't it be any related to permission for auto-ecearth project?

I don't think so, since this is a local git Hook...

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Mar 1, 2022, 10:18

Hello @eferre1 

For some reason today is working, I only did a new experiment from scratch

```
The project folder /home/dbeltran/autosubmit/a00o/proj has been created.
Cloning 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git into /home/dbeltran/autosubmit/a00o/proj
Clone command:  git clone --single-branch -b 1378-githooks_setup https://earth.bsc.es/gitlab/es/auto-ecearth3.git auto-ecearth3;
Cloning into 'auto-ecearth3'...
remote: Enumerating objects: 19673, done.
remote: Counting objects: 100% (5932/5932), done.
remote: Compressing objects: 100% (1182/1182), done.
remote: Total 19673 (delta 5620), reused 4774 (delta 4746), pack-reused 13741
Receiving objects: 100% (19673/19673), 18.93 MiB | 867.00 KiB/s, done.
Resolving deltas: 100% (13205/13205), done.
Githook + Checkout and Submodules: cd /home/dbeltran/autosubmit/a00o/proj/auto-ecearth3;  git config core.hooksPath ./.githooks ; git checkout 6dfaaf1e140543bdc7dfc1c570e2feae4f4bc6e1; git submodule init; git submodule update test_cases; 
Note: switching to '6dfaaf1e140543bdc7dfc1c570e2feae4f4bc6e1'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by switching back to a branch.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -c with the switch command. Example:

  git switch -c <new-branch-name>

Or undo this operation with:

  git switch -

Turn off this advice by setting config variable advice.detachedHead to false

HEAD is now at 6dfaaf1e Modify how the post-checkout git hook is disabled to be able to share other hooks
Running post-checkout githook to configure the SVN repositories for sources
Submodule 'sources' (https://earth.bsc.es/gitlab/svn/ecearth-mirror.git) registered for path 'sources'
Cloning into '/home/dbeltran/autosubmit/a00o/proj/auto-ecearth3/sources'...
```

It mentions:

> modify how the post-checkout git hook is disabled to be able to share other hooks

Not sure if it is related

**Comment by kinow** [comment]
In GitLab by @eferre1 on Mar 1, 2022, 10:28

Good news, thanks @dbeltrankyl.

That mention is the message of the commit of which you did the checkout (6dfaaf1e), the next lines are the ones of the githook that are running:

```
echo "Running post-checkout githook to configure the SVN repositories for sources"

git submodule init sources/
git submodule update sources/
cd sources/
set_svn
cd ..
mv .githooks/post-checkout .githooks/post-checkout_completed

echo "SVN repository set up, post-checkout githook completed"
```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 12, 2022, 12:41

Hi @dbeltrankyl, I tested this with the 3.14.0b autosubmit version but it is still not working as expected, I have done a few tries and from the looks of it, it is an issue with the autosubmit machines not triggering the `hooks`, even when running the `git checkout` manually.

I tried it from the workstation, AS01 and AS02 (from the same experiment folder after a refresh):

```
eferre1@ bscesautosubmit02:/esarchive/autosubmit/a4h4/proj/auto-ecearth3> git checkout 
Your branch is up-to-date with 'origin/1378-githooks_setup'.


eferre1@ bscesautosubmit01:/esarchive/autosubmit/a4h4/proj/auto-ecearth3> git checkout 
Your branch is up-to-date with 'origin/1378-githooks_setup'.


eferre1@ bscearth358:/esarchive/autosubmit/a4h4/proj/auto-ecearth3> git checkout 
Your branch is up-to-date with 'origin/1378-githooks_setup'.
Running post-checkout githook to configure the SVN repositories for sources
Cloning into '/esarchive/autosubmit/a4h4/proj/auto-ecearth3/sources'...
(···)
```
And as you can see, from the autosubmit machines it is not triggering the hook, even if it is completely set up (and from the workstation it triggers it). So it might not be an issue with the autosubmit procedure anymore, but it is still some weird reason making it not work (and my bashrc is empty, so nothing that might be affecting it).

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 12, 2022, 12:57

Could it be the GIT version?

```
dbeltran@ bscearth357:~> git --version
git version 2.13.6
dbeltran@ bscearth357:~> ssh autosubmit
Last login: Tue Jul 12 12:55:39 2022 from bscearth357.int.bsc.es
Have a lot of fun...
dbeltran@ bscesautosubmit01:~> git --version
git version 2.1.4
```

We can load 

```
------------------------- /shared/earth/modules/tools --------------------------
   git/2.6.1-foss-2015a    git/2.20.1-foss-2015a (D)    git-lfs/2.6.1
```

Maybe we could try to load 2.20 in both machines for test

At home, where it also works, I'm using

```
dbeltran@ bsces107894: git --version
git version 2.25.1
```

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 12, 2022, 13:00

Yes, it seems to be that, I just tried in both machines to load the 2.20 version and it worked:

```
eferre1@ bscesautosubmit01:/esarchive/autosubmit/a4h4/proj/auto-ecearth3> module load git/2.20.1-foss-2015a 
eferre1@ bscesautosubmit01:/esarchive/autosubmit/a4h4/proj/auto-ecearth3> git checkout 
Your branch is up to date with 'origin/1378-githooks_setup'.
Running post-checkout githook to configure the SVN repositories for sources
Cloning into '/esarchive/autosubmit/a4h4/proj/auto-ecearth3/sources'...
Username for 'https://earth.bsc.es': ^C
```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 12, 2022, 13:03

So then it is a matter of telling to users to load that git version.. FYI: @mcastril

**Comment by kinow** [comment]
In GitLab by @eferre1 on Jul 12, 2022, 13:05

Yes, I can confirm also that loading the 2.20 version and then running the refresh it works and sets up the svn repositories as expected.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Jul 21, 2022, 10:39

Seems like we can close this issue for AS side, the issue is resolved when git 2.20 is loaded

**Comment by kinow** [comment]
In GitLab by @eferre1 on Nov 8, 2022, 11:37

Hi @dbeltrankyl, following the failure of the clone on auto-ecearth3 on https://earth.bsc.es/gitlab/es/autosubmit/-/issues/667#note_184614, we would need a couple changes here to make sure that it doesn't fail while running the githook:

- From AS (@dbeltrankyl): 
  - to check what version of git has the user loaded and if it isn't an updated one (right now `2.20.1` is the one loaded after `module load git` in both AS machines) don't do the `git config core.hooksPath` so the hooks aren't deployed and activated later by accident.
  - if it is possible that at the `create`/`refresh` commands, it parses the files in the `.githooks` folder, so that the new `PROJECT_SUBMODULES_DEPTH` can be used there when running the hook (since it is running a `submodule update`).
- From auto-ecearth3 (myself @eferre1):
  - if the second part of AS is feasible, modify the `post-checkout` githook to take the depth into account and have a default value for when it is empty or the hooks manually set and activated.
  - modify the development wiki page to reflect this changes as it affects developers that need to modify the submodules.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 8, 2022, 11:54

Hello @eferre1,

Thanks for the notes, 

> to check what version of git the user has loaded and if it isn't an updated one (right now `2.20.1` is the one loaded after `module load git` in both AS machines) don't do the `git config core.hooksPath` so the hooks

Do you know which is the minimum version in which the git hook works? When used outside office machines it should also work.

> if it is possible that at the `create`/`refresh` commands, it parses the files in the `.githooks` folder, so that the new `PROJECT_SUBMODULES_DEPTH` can be used there when running the hook (since it is running a `submodule update`).

What do you mean by parse? what should AS parse? Do you mean that AS should detect if .githook exists?

**Comment by kinow** [comment]
In GitLab by @eferre1 on Nov 8, 2022, 12:15

> Do you know which is the minimum version in which the git hook works?

I'm not sure which is the exact version that works, since not all of them are available at the workstations, but it doesn't work for `git/2.6.1-foss-2015a` and it works with `2.13.6`(default one in the workstations) and `git/2.20.1-foss-2015a`. I'm looking if it is written somewhere in the internet documentation of git, but didn't find it yet.

> What do you mean by parse? what should AS parse? Do you mean that AS should detect if .githook exists?

The idea would be to use the `expdef` variable in the githook, the same way that autosubmit parses the templates to run them. For example, if we change the file to `post-checkout.tmpl`, then AS would be able to create the `post-checkout` file from it substituting the `%PROJECT_SUBMODULES_DEPTH%` by the value in the config file? Or something similar to use the variable defined inside the githook file.

**Comment by kinow** [comment]
In GitLab by @eferre1 on Nov 8, 2022, 12:52

According to this webpage ([devopedia](https://devopedia.org/git-hooks)), the githooks have been available always (from the `.git/hooks` path), but the path configuration with `git config core.hooksPath PATH` only since the version 2.9.0

**Comment by kinow** [comment]
In GitLab by @mcastril on Nov 9, 2022, 13:16

> The idea would be to use the `expdef` variable in the githook, the same way that autosubmit parses the templates to run them. For example, if we change the file to `post-checkout.tmpl`, then AS would be able to create the `post-checkout` file from it substituting the `%PROJECT_SUBMODULES_DEPTH%` by the value in the config file? Or something similar to use the variable defined inside the githook file.

Yes, the idea would be that Autosubmit parses the githooks the same way that it parses the template to generate bash scripts. Do you think that it's possible @dbeltrankyl ?

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 9, 2022, 13:21

Yes, it is possible is similar to what I did in AS4 for the custom_config parameter

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 15, 2022, 16:27

Hello @eferre1 , 

I've added the .githooks parser. It works for all files under .githooks and it is performed automatically before the checkout.

**Comment by kinow** [comment]
In GitLab by @eferre1 on Nov 15, 2022, 16:31

Ok, I'll edit the file then to use the new `%PROJECT_SUBMODULES_DEPTH%` variable from the `expdef` config file to take the depth into account.
Thanks.

**Comment by kinow** [comment]
In GitLab by @mcastril on Nov 17, 2022, 10:12

If now we can parse Autosubmit variables in the Git Hooks, maybe we can only do this operation in non-production experiments

**Comment by kinow** [comment]
In GitLab by @eferre1 on Nov 17, 2022, 10:47

Yes, I'll re-open and keep working on the auto-ecearth3 side on https://earth.bsc.es/gitlab/es/auto-ecearth3/-/issues/1378. From the AS part would remain the part about checking the git version if you think it is feasible @mcastril, @dbeltrankyl.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 17, 2022, 10:48

Oh, I forgot about the git version part. I'll check

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 23, 2022, 12:25

It should've available now 8174f4ed7967c136fb6001d2c9afe66d793ed1a7

**Comment by kinow** [comment]
In GitLab by @eferre1 on Nov 24, 2022, 08:16

Thanks @dbeltrankyl, I'll test it. Another question about the parsing of the `.githooks` folder, can the behavior be changed to parse the files like it is done with the templates (but creating the new file on the same `.githooks` folder)? 

`post-checkout.tmpl` will be parsed into `post-checkout` for example. That way we can avoid tracking any changes to the files and users committing them by mistake, otherwise, the parsed `post-checkout` file is detected as modified and some user might commit the change by error:

```
eferre1@ bscearth358:/esarchive/autosubmit/a3p0/proj/auto-ecearth3/.githooks> git status
On branch 1378-update_githook
Your branch is up-to-date with 'origin/1378-update_githook'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   post-checkout
```

And here is the diff:

```diff
diff --git a/.githooks/post-checkout b/.githooks/post-checkout
index 375a69ee..c09bef14 100755
--- a/.githooks/post-checkout
+++ b/.githooks/post-checkout
@@ -26,7 +26,7 @@ function track_svn_branches()
   git fetch -v --depth=${PROJECT_SUBMODULES_DEPTH}
 }
 echo ${PRODUCTION_EXP}
-PRODUCTION_EXP=%PRODUCTION_EXP%
+PRODUCTION_EXP=FALSE
 echo ${PRODUCTION_EXP}
 if [[ (${PRODUCTION_EXP:0:1} == "'Reference to deleted milestone   &&  ${PRODUCTION_EXP: -1:1} == ''Reference to deleted milestone ) ||  ${PRODUCTION_EXP} == '" ]] ; then
   PRODUCTION_EXP=TRUE
@@ -38,7 +38,7 @@ if [[ ! -f ".githooks/post-checkout_completed" ]] ; then
     touch .githooks/post-checkout_completed
   else
     echo ${PROJECT_SUBMODULES_DEPTH}
-    PROJECT_SUBMODULES_DEPTH=%PROJECT_SUBMODULES_DEPTH%
+    PROJECT_SUBMODULES_DEPTH=50
     echo ${PROJECT_SUBMODULES_DEPTH}

```

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 24, 2022, 08:19

Oh, I see I can do it, but I would need a common .ext to differentiate the files, do you think .tmpl is good enough?

**Comment by kinow** [comment]
In GitLab by @eferre1 on Nov 24, 2022, 09:32

Sure, the hooks are activated by name, so anything that is not the plane name (any extension) will not trigger it and can be used to differentiate them, so AS only needs to parse it and create the new file without the `.tmpl` from the file name. That would be exactly what we need.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Nov 24, 2022, 09:46

Hello, 

I'm updating the 3.14.0b, it will finish in 5~min. (Updated)

 * Now AS will only parser the files under .githooks that has .tmpl as an extension.
 * Then it will create the file with 750 permissions under .githooks and without extension

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 5, 2022, 15:23

I think there were no pending tasks for this one, no?

**Comment by kinow** [comment]
In GitLab by @eferre1 on Dec 5, 2022, 15:38

I don't think so. Thanks for all the updates @dbeltrankyl.

**Comment by kinow** [comment]
In GitLab by @dbeltrankyl on Dec 5, 2022, 15:41

thanks, 

Great! closing it then

**Comment by kinow** [comment]
In GitLab by @ainagaya on Mar 7, 2024, 11:05

mentioned in merge request digital-twins/de_340/workflow!202
